STYLEGUIDE.md

```diff
@@ -0,0 +1,376 @@
+for separation. For example:
+
+```
+int const kDaysInAWeek = 7;
+int const kAndroid8_0_0 = 24;  // Android 8.0.0
+```
+
+All such variables with static storage duration (i.e., statics and globals, see
+Storage Duration for details) should be named this way. This convention is
+optional for variables of other storage classes, e.g., automatic variables,
+otherwise the usual variable naming rules apply.
+
+#### Function Names
+
+Regular functions have mixed case; accessors and mutators may be named like
+variables.
+
+Ordinarily, functions should start with a non-capital letter and have a capital
+letter for each new word.
+
+```
+addTableEntry()
+deleteUrl()
+openFileOrDie()
+```
+
+(The same naming rule applies to class- and namespace-scope constants that are
+exposed as part of an API and that are intended to look like functions, because
+the fact that they're objects rather than functions is an unimportant
+implementation detail.)
+
+Accessors and mutators (get and set functions) may be named like variables.
+These often correspond to actual member variables, but this is not required. For
+example, `int count()` and void `setCount(int count)`.
+
+#### Namespace Names
+
+Namespace names are all lower-case, with words separated by underscores.
+Top-level namespace names are based on the project name . Avoid collisions
+between nested namespaces and well-known top-level namespaces.
+
+The code in that namespace should usually be in a directory whose basename
+matches the namespace name (or in subdirectories thereof).
+
+Avoid nested namespaces that match well-known top-level namespaces. Collisions
+between namespace names can lead to surprising build breaks because of name
+lookup rules. In particular, do not create any nested std namespaces. Prefer
+unique project identifiers (`websearch::index`, `websearch::index_util`) over
+collision-prone names like `websearch::util`. Also avoid overly deep nesting
+namespaces ([TotW #130](https://abseil.io/tips/130)).
+
+For `internal` namespaces, be wary of other code being added to the same `internal`
+namespace causing a collision (internal helpers within a team tend to be related
+and may lead to collisions). In such a situation, using the filename to make a
+unique internal name is helpful (`websearch::index::frobber_internal` for use in
+`frobber.h`).
+
+#### Enumerator Names
+
+Enumerators (for both scoped and unscoped enums) should be named like constants,
+not like macros. That is, use `kEnumName` not `ENUM_NAME`.
+
+```
+enum class UrlTableError {
+  kOk = 0,
+  kOutOfMemory,
+  kMalformedInput,
+};
+```
+```
+// BAD practice
+enum class AlternateUrlTableError {
+  OK = 0,
+  OUT_OF_MEMORY = 1,
+  MALFORMED_INPUT = 2,
+};
+```
+
+Until January 2009, the style was to name enum values like macros. This caused
+problems with name collisions between enum values and macros. Hence, the change
+to prefer constant-style naming was put in place. New code should use
+constant-style naming.
```

Comment:
This probably refers to Google's situation but not to ours.
We also have some old code that does not follow the naming conventions listed here.
I think we should have a preamble section in "Naming" saying that all _new_ code should follow the naming conventions mentioned here, and that some old code may not follow it yet.